Change UrlDecode behavior and merge test cases for .Net core and desktop

schen381 opened the pull request:
1. Tested in .Net framework, confirmed the difference. Also confirmed the correctness of documentation
2. Merge the test for .Net core and desktop. Since now their behaviors match, no need to test separately (https://github.com/dotnet/corefx/pull/17618#issuecomment-289933351)

Fix #17627

CIPop commented on src/System.Web.HttpUtility/tests/HttpUtility/HttpUtilityTest.cs:
```diff
@@ -412,16 +412,10 @@ public void ParseQueryString_ToString()
                 new object[] {"http://127.0.0.1:8080/app%%Dir/page.aspx?foo=b%%r", "http://127.0.0.1:8080/app%%Dir/page.aspx?foo=b%%r"},
                 new object[] {"http://127.0.0.1:8080/appDir/page.aspx?foo=ba%r", "http://127.0.0.1:8080/appDir/page.aspx?foo=b%61%r"},
                 new object[] {"http://127.0.0.1:8080/appDir/page.aspx?foo=b%uu0061r", "http://127.0.0.1:8080/appDir/page.aspx?foo=b%uu0061r"},
+                new object[] {"http://127.0.0.1:8080/appDir/page.aspx?foo=b%u0061r", "http://127.0.0.1:8080/appDir/page.aspx?foo=b%u0061r"},
```
@schen381 Please check if there are any more .Net Framework tests that we could port to CoreFX.

schen381 replied:
sure, I will look into .Net Framework source code.

jkotas commented on src/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs:
```diff
@@ -283,45 +283,14 @@ internal static byte[] UrlDecode(byte[] bytes, int offset, int count)
                 {
-                    if (bytes[pos + 1] == 'u' && i < count - 5)
-                    {
-                        int h1 = HttpEncoderUtility.HexToInt((char)bytes[pos + 2]);
```
.NET Framework implementation does have this piece of code that is being remove here: https://github.com/Microsoft/referencesource/blob/4fe4349175f4c5091d972a7e56ea12012f1e7170/System.Web/Util/HttpEncoder.cs#L519

It does not sound right to remove this to match .NET Framework.

schen381 replied:
.Net core also has same piece of code here: https://github.com/dotnet/corefx/blob/master/src/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs#L341 and that's not this PR's concern
Note the function signature difference, the function I modified is UrlDecode(byte[] bytes, int offset, int count), not UrlDecode(byte[] bytes, int offset, int count, Encoding encoding)

stephentoub replied:
If this code being removed didn't exist in desktop, it must have been explicitly added to core for some reason, right? Was it fixing a bug (I.e. the condition called out in this issue)? And if so, why are we removing it? Seems like the right fix would be to fix this  eventually in desktop and in the meantime leave the test disabled for desktop?

stephentoub replied:
cc: @alexperovich

alexperovich replied:
This file was a direct copy from .NET Framework, I don't see how removing anything can make this match the implementation there.

stephentoub replied:
> This file was a direct copy from .NET Framework

@alexperovich, do you know how this:
https://referencesource.microsoft.com/#System.Web/Util/HttpEncoder.cs,477
ended up differing from this:
https://github.com/dotnet/corefx/blob/3bbc5ddc7006bafa0d68883dde90fbd0f839d57c/src/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs#L284-L291
?

alexperovich replied:
I do not. I ported the code from the internal branch not from reference source so it could have been changed recently.

schen381 replied:
>If this code being removed didn't exist in desktop, it must have been explicitly added to core for some reason, right? Was it fixing a bug (I.e. the condition called out in this issue)?

@stephentoub  I have looked into NetFXDev1, but NCL doesn't own the code, and I can't find any information from our side

schen381 commented:
@CIPop I have looked into NetFXDev1, http://index, Microsoft/referencesource repo and cannot find any .Net Framework tests for this method. I can add more test cases for the method by myself, but the problem is I'm not sure whether this modification to match desktop behavior is needed or not, because the documentation only talks about characters which are not allowed in a URL. I agree with David's understanding here: https://github.com/dotnet/corefx/issues/17627#issuecomment-292601670

 @alexperovich Could you please provide some instructions, like how to see the content of internal branch you used to port the code? I want to check its commit history and see if there are more test cases there.

alexperovich commented:
IIRC I ported this from NetFXDev1, I can't find any code there where %u is handled though.

I think this change is the correct direction though. The %u0000 encoding isn't in any RFC for Uris, though the spec for JavaScript allows for it. Matching the desktop behavior in this case is correct.

CIPop commented:
@Tratcher @danroth27 PTAL 

I think this type is actually something originating in ASP.Net (from the `System.Web` namespace) so you may have some context.
Also, please let us know if you are aware of any additional tests that we can port from any internal or OSS sources to verify that the behavior is the same in .Net Framework and .Net Core.

danmosemsft commented:
@Tratcher @danroth27 PTAL

schen381 commented:
Had offline discussion with @alexperovich . Here is my report:
1.	Source code was ported from NetFXDev1, and there was no test for the UrlDecode() and UrlDecodeToBytes()
2.	Test cases were ported from Mono
In order to pass the test cases, Alex took the implementation from the other overload of UrlDecode() and updated the code because the test was failing. (Mono has different implementation for the methods)

The reason why our CI didn’t report the test failure when he ported the code into .Net core is that, test cases were put in a wrong namespace. Later namespace was corrected, and the issue occurred: https://github.com/dotnet/corefx/commit/c740fae63e0363b8eca1ee1f34d45732f30cc50c#diff-d281dfb17a5b66587fda59f009cc6ce0L12

Here are details:
Mono’s UrlDecodeToBytes has its own implementation: https://github.com/mono/mono/blob/0af1ea70c29a66a59765183b45d6b75fa6c3d8fa/mcs/class/System.Web/System.Web/HttpUtility.cs#L303 
In .Net framework, UrlDecodeToBytes will use UrlDecode’s implementation (line 1503): http://index/?query=UrlDecodeToBytes&rightProject=System.Web&file=httpserverutility.cs&line=1503

Summary of my change:
1.	Matched framework behavior (deleted the code from overload method)
2.     Corrected wrong test data, and merge test cases

cc: @stephentoub @davidsh @CIPop

karelz commented:
Triage: @CIPop please push it through ... looks like the code copy concern was addressed above.
